### PR TITLE
feat: add support for SparkML KMeansModel conversion

### PR DESCRIPTION
**Pull request by memoryz**
#538

### CODE REVIEW

**Comment by memoryz** [comment]
@xadupre This is ready for review.

File: onnxmltools/convert/sparkml/operator_converters/k_means.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            "alpha": -2.0,
+            "beta": 1.0,
+            "transB": 1,
+        }
+        container.add_node(
+            op_type="Gemm", 
+            inputs=[operator.inputs[0].full_name, centers_variable_name, input_row_squared_sum_variable_name], 
+            outputs=[gemm_output_variable_name],
+            op_domain="ai.onnx", 
+            op_version=13,
```
Comment (xadupre) [inline onnxmltools/convert/sparkml/operator_converters/k_means.py:71]: Same comment as above.

File: tests/sparkml/test_k_means.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            model, 
+            'Sparkml KMeansModel', 
+            [('features_euclidean', FloatTensorType([None, 3])), ('features_cosine', FloatTensorType([None, 3]))],
+            target_opset=TARGET_OPSET
+            )
+
+        self.assertTrue(model_onnx is not None)
+        self.assertTrue(model_onnx.graph.node is not None)
+        
+        with open('kmeans.onnx', 'wb') as f:
```
Comment (xadupre) [inline tests/sparkml/test_k_means.py:52]: do you need that?

Comment (memoryz) [reply tests/sparkml/test_k_means.py]: I missed that. Thanks.

File: onnxmltools/convert/sparkml/operator_converters/k_means.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        )
+        
+        # Euclidean Distance Squared = input_row_squared_sum - 2 * input * Transpose(Center) + Transpose(centers_row_squared_sum)
+        # [N x K]
+        container.add_node(
+            op_type="Add",
+            inputs=[gemm_output_variable_name, centers_row_squared_sum_variable_name],
+            outputs=[distance_output_variable_name],
+            op_domain="ai.onnx",
+            op_version=13,
```
[the remark this answers is not in the document]
Comment (memoryz) [reply onnxmltools/convert/sparkml/operator_converters/k_means.py]: Done.

**Comment by memoryz** [comment]
@xadupre ready for another round of code review.

File: onnxmltools/convert/sparkml/operator_converters/k_means.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            **gemm_attrs
+        )
+
+        # Cosine similarity = gemm_output / input_row_norm2 / centers_row_norm2: [N x K]
+        div_output_variable_name = scope.get_unique_variable_name("div_output")
+        container.add_node(
+            op_type="Div",
+            inputs=[gemm_output_variable_name, input_row_norm2_variable_name],
+            outputs=[div_output_variable_name],
+            op_version=7, # Setting to version 7 for broadcasting support
```
Comment (xadupre) [inline onnxmltools/convert/sparkml/operator_converters/k_means.py:130]: Maybe you can remove the opset but raise an exception when the requested opset is below 7. That's usually what we do.

Comment (memoryz) [reply onnxmltools/convert/sparkml/operator_converters/k_means.py]: Done.

**Comment by memoryz** [comment]
@xadupre ready for another round of code review. :)

**Comment by memoryz** [comment]
@xadupre ping.

**Review by xadupre — APPROVED** [review]

**Pull request merged**